### ISSUE TITLE
APIError Exceptions / Errors makes no sense

### ISSUE THREAD

**Issue by skela** [dropbox/dropbox-sdk-python#35]
Sorry to raise this issue, but I've given them a good try, and still can't make sense of them.
I've checked out your examples, and I've read your documentation:
http://dropbox-sdk-python.readthedocs.org/en/master/moduledoc.html#dropbox.files.ListFolderError

I think you need a lot more examples of usual errors that might crop up whilst using your API.

Since I'm here anyway, can anyone point me in the right direction on how to handle the following scenario:
I want to check if a folder exists, and if it is not found, I want to create that folder.

Other than that, its pretty good stuff!

**Comment by smarx** [comment]
For your particular scenario, you may be able to just do nothing. Folders are implicitly created in Dropbox, so if you add a file at "/foo/bar", the folder "/foo" will be created if it doesn't already exist.

For your direct question, you probably want to call `.is_path()` on the error to confirm that you got back a `LookupError`. You can then call `.get_path()` to get the actual LookupError and test what the issue was. (In particular, you're probably looking for a `not_found`. I find the HTTP documentation helpful here (expand `LookupError` to see the different cases): https://www.dropbox.com/developers/documentation/http/documentation#documentation-files-list_folder.

**Comment by skela** [comment]
Thanks, good to know that its implicit!

Although I probably won't need this bit anymore, out of curiosity and in case I need to do something similar, or to handle another error, here's my snippet:

```
try:
	self.dbx.files_list_folder(self.db_remote_root)
except DBExceptions.ApiError as err:
	print '*** DBAPI error %s' % str(err.error.get_path())
	if err.error.is_path() and err.error.get_path()._tag == 'not_found':
		self.dbx.files_create_folder(self.db_remote_root)
except requests.ConnectionError:
	exit("Failed to connect to dropbox")
```

This works, but I'm accessing a protected member by reading the _tag.
What's the recommended way to handle an error here.

**Comment by braincore** [comment]
Try this:
```
if err.error.is_path() and err.error.get_path().is_not_found():
    ...
```

From the [`ListFolderError` doc](http://dropbox-sdk-python.readthedocs.org/en/master/moduledoc.html#dropbox.files.ListFolderError) you can see that `get_path()` returns a `LookupError`. From the [`LookupError` doc](http://dropbox-sdk-python.readthedocs.org/en/master/moduledoc.html#dropbox.files.LookupError), you can see the `is_*` methods including `is_not_found()`.

**Comment by smarx** [comment]
I haven't tried it, but probably `err.error.get_path().is_not_found()`?

**Comment by braincore** [comment]
@smarx beatcha! :)

@skela I'm going to leave this thread open until we clarify error handling further (whether by examples or docs).

**Comment by skela** [comment]
@braincore and @smarx thanks, that works. And yeah, good idea about keeping this open, I'm probably not the only one who got / will get lost here :D

**Comment by vinodc** [comment]
Is there a way to handle errors with some common logic rather than explicitly checking for every single type of error?

I'm hoping to include the error handling logic in a decorator so that our code doesn't get too verbose for the multitude of API requests it makes. When an error occurs, I'd like to perform some common actions, such as log necessary details and obtain a message to display to the user. However, the SDK makes this difficult because the error object may/may not have several different types of _tag objects, each with their own attributes, and each requiring special introspection if I wanted to generally handle any kind of error.

I was hoping ApiError would provide some common abstraction, but all it provides is the error attribute so that isn't very helpful. This feels far more complicated that the v1 API.

**Comment by braincore** [comment]
@vinodc We have many routes that are completely unrelated, so the amount of handling that can be shared is naturally limited.

If you're just looking to do some logging, I recommend calling `repr(err)` where `err` is an `ApiError`. It will return a string that describes the entire error tree.

You can also use the `user_message` attribute of `ApiError` which is sometimes populated with a localized message that can be displayed directly to the user.

Finally, if you do want some common error handling logic, I suggest you focus on specific ones. For example, we can handle all `LookupError`s which is common across listing a folder, getting metadata, downloading, searching, deleting, and more:

```
from contextlib import contextmanager
from dropbox.exceptions import ApiError
from dropbox.files import LookupError

@contextmanager
def error_handler_for_file_methods():
    try:
        yield
    except ApiError as e:
        if hasattr(e.error, 'is_path') and e.is_path():
            # Sanity check that we're getting the type we think we are
            assert isinstance(e.get_path(), LookupError)
            # Do something with the LookupError
```

Let us know what approach you end up taking.

**Comment by vinodc** [comment]
Thanks @braincore . I ended up doing something similar and caught DropboxException errors before introspecting each exception for its type and attributes so I could handle the common cases like LookupError and WriteError. Using this method is a bit verbose, but the presence of `_value` and `user_message_text` (sometimes populated, as you mentioned) does help cut down on the number of lines. I'm still ignoring other errors, and have a generic catch-all for them, which isn't ideal.

For HttpErrors though, it is helpful that the subclasses end up with a common attribute `status_code`. I've extracted a message with `message = getattr(exception, 'message', getattr(exception, 'error', ''))` as well, so this case is more manageable than ApiError.

One other thing that I found odd but isn't really an issue is that using an invalid token results in a BadInputError rather than an AuthError.
